pkg/geo/geomfn/reverse.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/twpayne/go-geom"
+)
+
+// Reverse returns a modified geometry by reversing the order of its vertexes
+func Reverse(geometry *geo.Geometry) (*geo.Geometry, error) {
+	g, err := geometry.AsGeomT()
+	if err != nil {
+		return nil, err
+	}
+
+	g, err = reverse(g)
+	if err != nil {
+		return nil, err
+	}
+
+	return geo.NewGeometryFromGeomT(g)
+}
+
+func reverse(g geom.T) (geom.T, error) {
+	if geomCollection, ok := g.(*geom.GeometryCollection); ok {
+		return reverseCollection(geomCollection)
+	}
+
+	switch t := g.(type) {
+	case *geom.Point, *geom.MultiPoint: // cases where reverse does change the order
+		return g, nil
+	case *geom.LineString:
+		g = geom.NewLineStringFlat(t.Layout(), reverseCoords(g.FlatCoords(), g.Stride())).SetSRID(g.SRID())
+	case *geom.Polygon:
+		g = geom.NewPolygonFlat(t.Layout(), reverseCoords(g.FlatCoords(), g.Stride()), t.Ends()).SetSRID(g.SRID())
+	case *geom.MultiLineString:
+		g = geom.NewMultiLineStringFlat(t.Layout(), reverseMulti(g, t.Ends()), t.Ends()).SetSRID(g.SRID())
+	case *geom.MultiPolygon:
+		var ends []int
+		for _, e := range t.Endss() {
+			ends = append(ends, e...)
+		}
+		g = geom.NewMultiPolygonFlat(t.Layout(), reverseMulti(g, ends), t.Endss()).SetSRID(g.SRID())
+
+	default:
+		return nil, geom.ErrUnsupportedType{Value: g}
+	}
+
+	return g, nil
+}
+
+func reverseCoords(coords []float64, stride int) []float64 {
+	n := len(coords)
+
+	for i := 0; i < n/2; i += stride {
+		left := coords[i : i+stride]
+		right := coords[n-i-stride : n-i]
+
+		tmp := append([]float64(nil), left...)
```

Comment:
can we do this in place, i.e. without allocating a new array?
```
for i := 0; i < len(coords); i += stride {
   for j = 0; j < stride; j++ {
      coords[i+j], coords[len(coords)-i+j] = coords[len(coords)-i+j], coords[i+j]
    }
}
```

Reply:
Yep, nice suggestion. I had to tweak your code a bit though:
```go
	for i := 0; i < len(coords)/2; i += stride {
		for j := 0; j < stride; j++ {
			coords[i+j], coords[len(coords)-stride-i+j] = coords[len(coords)-stride-i+j], coords[i+j]
		}
	}

```